### PR TITLE
Adding TE comm-gemm overlap to llama models

### PR DESCRIPTION
**Pull request by abhinavgoel95**
# Description

Adding TE's Comm GeMM overlap API to llama2 models.

The rest of the description includes relevant details and context, examples:
Improves TP performance by overlapping communication and compute.

# Tests

```
RUN_CMD="python -m MaxText.train /opt/workspace/te-maxtext/src/MaxText/configs/base.yml run_name=logdir use_iota_embed=true scan_layers=True    steps=31 per_device_batch_size=0.5 model_name=${MODEL} remat_policy=minimal_flash enable_checkpointing=false logits_dot_in_fp32=false    base_output_directory=/opt/maxtext/local_train dataset_path=local dataset_type=synthetic attention=cudnn_flash_te max_target_length=8192    enable_goodput_recording=false monitor_goodput=false quantization="" tokenizer_path=tiktoken hardware=gpu    dcn_fsdp_parallelism=1 ici_data_parallelism=1    ici_fsdp_parallelism=2 dcn_data_parallelism=1    ici_tensor_sequence_parallelism=4 dcn_tensor_parallelism=1 ${PROFILING_SETTINGS} use_te_comm_gemm_overlap=true shardy=false"
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x ] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x ] I have run end-to-end tests tests and provided workload links above if applicable.
- [x ] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/MaxText/layers/llama2.py [hunk 1]
```diff
@@ -120,6 +131,8 @@ def __init__(
         model_mode=model_mode,
         rngs=rngs,
         use_pre_norm=use_te,
+        use_te_comm_gemm_overlap=config.use_te_comm_gemm_overlap,
```
Comment (jberchtold-nvidia) [inline src/MaxText/layers/llama2.py]: `use_te_comm_gemm_overlap` doesn't need to be an explicit arg here, we pass in the full `config` to the MlpBlock so we could check `config.use_te_comm_gemm_overlap` internal to the MlpBlock of DenseGeneral and get support for all model types, not just llama2, right? Lmk if I'm missing something, thanks!

File: src/MaxText/train.py [hunk 2]
```diff
@@ -514,6 +514,7 @@ def transformer_engine_context():
     mesh_resource = MeshResource(
       dp_resource = "data",
       tp_resource = "tensor",
+      tpsp_resource = "tensor_sequence",
```
Comment (jberchtold-nvidia) [inline src/MaxText/train.py:517]: Good catch, thanks!

File: src/MaxText/layers/llama2.py [hunk 3]
```diff
@@ -164,7 +177,11 @@ def __call__(
     attention_lnx = nn.with_logical_constraint(attention_lnx, self.activation_axis_names)
     intermediate_inputs = inputs + attention_lnx
 
-    use_te = self.config.quantization.startswith("te_")
+    if self.config.quantization is None:
```
Comment (jberchtold-nvidia) [inline src/MaxText/layers/llama2.py:179]: Thanks for fixing this!

File: src/MaxText/layers/llama2.py (hunk 1 above)
Comment (jberchtold-nvidia) [reply src/MaxText/layers/llama2.py]: Same for `collection_op_sets`, won't that be true for all MlpBlocks?

File: src/MaxText/layers/linears.py
Comment (jberchtold-nvidia) [inline src/MaxText/layers/linears.py:514]: I think this codepath will skip Comm+Gemm overlap

File: src/MaxText/layers/linears.py [hunk 4]
```diff
@@ -212,6 +233,17 @@ def __call__(self, inputs: Array, _initializing: bool = False) -> Array:
       kernel = jnp.asarray(kernel, self.dtype)
 
     contract_ind = tuple(range(0, len(self.axis)))
+
+    if self.use_te_comm_gemm_overlap:
+      output = te_dense(
```
Comment (jberchtold-nvidia) [inline src/MaxText/layers/linears.py]: This will prevent TE (or any) quantization from being applied as the quantization is applied via `_compute_dot_general_nnx` below. There is no global fp8_autocast context here to allow us to call `te_dense` with automatic quantization. The TE quant integration logic is internal to MaxText's `quantizations.py` and TE provides a `dot_general_cls` for quantization.

Comment (jberchtold-nvidia) [inline src/MaxText/layers/linears.py]: We have a few competing codepaths here that will make this potentially difficult to integrate properly in all cases

1. If TE quantization is enabled and `fused_mlp=False`. Quantization of `DenseGeneral`. When `fused_mlp=False`, TE can't do a full TE MlpBlock with fused act+quant, so we instead apply quantization directly to the DenseGenerals in the same way other MaxText quantizations work. This is invoked via `_compute_dot_general_nnx` below which ends up calling `dot_general_cls` for TE from `src/MaxText/maxtext/quantizations.py`
2. If TE quantization is enabled and `fused_mlp=True`, we create `self.te_ln_mlp` on the MaxText MlpBlock below and instead run ln+mlp with fused ln+quant and fused act+quant from TE.
3. If TE Comm+Gemm overlap is enabled. Inside DenseGeneral, we call `te_dense` instead of `_compute_dot_general_nnx` and pass in args to perform Comm+Gemm overlap.

From reading the code, I think item 3 will conflict with items 1 and 2, with one feature blocking the other's functionality.

There will still be some duplicate handling of logic, but I'm wondering if it instead makes sense to do the following:

A. TE quantizations in quantizations.py provides a new TE "quantization" type called "te_noscale", "te_gemm" or some similar name which supports running MaxText with TE bf16 GEMMs instead of JAX GEMMs via `--quantization=te_gemm`. It's not really a quantization, but it integrates nicely with the dot_general_cls logic already present.
B. Kwargs is added to the MaxText quantizations `dot_general_cls` interface to pass through additional args. TE's specific dot_general_cls impl will check for `collective_op_set` as a kwarg, and if so pass it through to it's own call of TE's `dense` VJP.

With A and B here we will support comm+gemm overlap without disabling quantization of DenseGeneral. This allows items 1 and 3 above to interop correctly. Currently TE only supports comm+gemm overlap for BF16, but in future we will want comm+gemm overlap for quantized dtypes as well.

C. For integrating with the TE LNMLP block (item 2 above), I think that can be done manually when it is created here https://github.com/nvjax-svc-0/maxtext/pull/26/files#diff-b39ed9bbf6af42fed33cc5552f8fa7e4cafb9d91373f1c7e6181e2f58967851bR408
This is a small bit of duplicate logic for passing collective_op_sets for multiple codepaths, but I think it will be fairly small and all features will operate correctly with each other.

Let me know what you think and if you have any questions. Thanks!

File: src/MaxText/configs/base.yml [hunk 5]
```diff
@@ -80,6 +80,8 @@ metrics_file: "" # for testing, local file that stores scalar metrics. If empty,
 # If true save metrics such as loss and TFLOPS to GCS in {base_output_directory}/{run_name}/metrics/
 gcs_metrics: False
 
+use_te_comm_gemm_overlap: False # use Transformer Engine's collective GeMM for tensor-parallel communication overlap
```
Comment (jberchtold-nvidia) [inline src/MaxText/configs/base.yml:83]: Thanks for making this PR!

File: src/MaxText/layers/linears.py (hunk 4 above)
Comment (abhinavgoel95) [reply src/MaxText/layers/linears.py]: Hmm! Thanks for the detailed response.

I understand. Right now, the comm-gemm overlap is only supported for BF16.

So maybe I'll move this code into quantizations.py and trigger it for BF16 using `--quantization=te_gemm`. That way it will be compatible with the existing TE infrastructure. 

Ill re-work this.

Comment (jberchtold-nvidia) [reply src/MaxText/layers/linears.py]: Thanks! I think Phuong put together a `--quantization=te_noscaling` (what I was calling `te_gemm`) for BF16 TE GEMMs in a PR here:
https://github.com/nvjax-svc-0/maxtext/pull/25

It has a lot of files changed that are unrelated, let me check with her if it's because the PR needs to be rebased. Hopefully that PR contains part of this re-working already

Comment (phu0ngng) [reply src/MaxText/layers/linears.py]: I closed PR #25 as it is no longer needed.

Comment (jberchtold-nvidia) [reply src/MaxText/layers/linears.py]: Checked with Phuong and that PR will be closed. But the relevant changes for `te_noscaling` are here and can be used as a starting point:
https://github.com/nvjax-svc-0/maxtext/pull/25/files#diff-3b2344e2f2f8a9cc5eebf885f5776ccf42d9981a3cb1d00bf99ef1a7e014ee86R753

In the same class in that file, the `dot_general_cls` method calls TE's `dense` VJP and is where you can pass thru the collective op set

File: src/MaxText/layers/linears.py
Comment (abhinavgoel95) [reply src/MaxText/layers/linears.py:514]: As discussed on the call, this will be addressed in a follow up PR

**Comment by abhinavgoel95** [comment]
@jberchtold-nvidia I have addressed all of your comments. Could you re-review. Thanks.

File: src/MaxText/layers/linears.py [hunk 6]
```diff
@@ -106,6 +123,8 @@ def __init__(
       parameter_memory_host_offload: bool = False,
       *,  # Following arguments are keyword-only
       rngs: nnx.Rngs = None,
+      use_te_comm_gemm_overlap: bool = False,
+      collective_op_set: CollectiveOpSet = None,
```
Comment (jberchtold-nvidia) [inline src/MaxText/layers/linears.py:127]: nit: We should probably omit the type annotation here and make a comment instead as I think this will give a type error if TE isn't available

File: src/MaxText/layers/linears.py [hunk 7]
```diff
@@ -410,6 +440,8 @@ def __init__(
             use_bias=self.use_bias,
             matmul_precision=self.config.matmul_precision,
             rngs=rngs,
+            use_te_comm_gemm_overlap=self.config.use_te_comm_gemm_overlap if idx==0 else False,
```
Comment (jberchtold-nvidia) [inline src/MaxText/layers/linears.py:443]: Why do we only do the first `wi` GEMM in the unfused MLP with comm+gemm overlap? I'm not familiar with the typical recipe here to know if this is a standard technique. Thanks!

Comment (jberchtold-nvidia) [reply src/MaxText/layers/linears.py:443]: Is this because in TE we only set up communication for two GEMMs in an MLP block since TE always does fused MLP?

File: src/MaxText/layers/llama2.py [hunk 8]
```diff
@@ -108,6 +112,13 @@ def __init__(
             rngs=rngs,
         )
 
+    if config.use_te_comm_gemm_overlap:
+        collective_op_set1 = CollectiveOpSet.create(forward_collective_op=CollectiveOp.ALL_GATHER)
```
Comment (jberchtold-nvidia) [inline src/MaxText/layers/llama2.py:116]: Is anything here specific to Llama2? Would it instead make sense to move this collective op set creation into the MlpBlock implementation itself where we can check `config.use_te_comm_gemm_overlap`? Then we would support all decoder block types that use the MlpBlock automatically. If you already mentioned the reasoning somewhere and I'm forgetting, apologies. Thanks!

File: src/MaxText/max_utils.py [hunk 9]
```diff
@@ -209,6 +209,23 @@ def maybe_initialize_jax_distributed_system(raw_keys):
...
+    from transformer_engine.jax.cpp_extensions.gemm import collective_gemm_bootstrap
+
+  except ImportError:
+    print("TransformerEngine Comm Gemm Overlap is not available")
+
+  if raw_keys["use_te_comm_gemm_overlap"]:
+    max_logging.log("Initializing TransformerEngine comm gemm overlap...")
+    collective_gemm_bootstrap(
+        num_total_devices=jax.process_count(),
+        num_devices_per_process=1,
```
Comment (jberchtold-nvidia) [inline src/MaxText/max_utils.py:225]: Since MaxText defaults to single process multi-GPU, should we add an assert here that `len(jax.devices()) == 1` to enforce that we're using only a single GPU per process here?

File: src/MaxText/layers/linears.py [hunk 10]
```diff
@@ -69,7 +86,7 @@ def _compute_dot_general(inputs, kernel, kernel_axes, axis, contract_ind, matmul
   dot_general = lax.dot_general
   matmul_precision = lax.Precision(matmul_precision)
   if quant:
-    dot_general_cls = quant.dot_general_cls(mesh_axes=kernel_axes)
+    dot_general_cls = quant.dot_general_cls(mesh_axes=kernel_axes, collective_op_set=self.collective_op_set)
```
Comment (jberchtold-nvidia) [inline src/MaxText/layers/linears.py:89]: Since this is not a method and instead a helper function, we don't have access to `self.collective_op_set` here, right? Should we instead pass `collective_op_set` as an additional arg

File: src/MaxText/layers/linears.py [hunk 11]
```diff
@@ -135,6 +154,8 @@ def __init__(
     self.use_bias = use_bias
     self.matmul_precision = matmul_precision
     self.parameter_memory_host_offload = parameter_memory_host_offload
+    self.use_te_comm_gemm_overlap = use_te_comm_gemm_overlap
```
Comment (jberchtold-nvidia) [inline src/MaxText/layers/linears.py:157]: I don't see `use_te_comm_gemm_overlap` used here. I think we can just pass thru `collective_op_set`, and that will be `None` if `config.use_te_comm_gemm_overlap` is `False`, right? If I missed something, lmk, thanks!
